Add openshift-pipelines-operator

nikhil-thomas opened the pull request:
Add an operator to install TektonCD Pipelines
This operator can run on both vanilla Kubernetes and on OpenShift

tested with
`operator-sdk scorecard`
and
`operator-courier verify`
on Apr 23 2019

sthaha commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+  - name: Tekton Pipeline GitHub Repository
+    url: https://github.com/tektoncd/pipeline
+  - name: Tekton Pipeline Operator GitHub Repository
+    url: https://github.com/tektoncd/pipeline
```
this needs to point to the actual operator repo

nikhil-thomas replied:
fixed
https://github.com/openshift/tektoncd-pipeline-operator

sthaha commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+          "apiVersion": "tekton.dev/v1alpha1",
+          "kind": "Install",
+          "metadata": {
+            "name": "example-install",
```
should this have a better name?

nikhil-thomas replied:
i was thinking that a placeholder name like **example-install** will be less distracting

SamiSousa replied:
I think the name of the example CR is fine in this case

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.package.yaml:
```diff
@@ -0,0 +1,5 @@
+packageName: tetektoncd-pipeline-operatorktoncd
+kind: PackageManifest
```
nit: `package.yaml` file does not use `kind` field, please remove it.

nikhil-thomas replied:
done

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.package.yaml:
```diff
@@ -0,0 +1,5 @@
+packageName: tetektoncd-pipeline-operatorktoncd
```
Please make sure your packageName matches the operator's directory, ie change the packageName to match the directory name, or vice versa.

nikhil-thomas replied:
this is an embarrassing mistake. Thank you for catching it :+1:

nikhil-thomas replied:
done

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+          "kind": "Install",
+          "metadata": {
+            "name": "example-install",
+            "namespace": "tekton-pipelines"
```
nit: Please remove this, there is no need to specify the namespace.

nikhil-thomas replied:
> @nikhil-thomas Thank you for submitting this operator! Is this operator's image built by the Red Hat build process?

i used `operator-sdk build`.
`registry.access.redhat.com/ubi8-dev-preview/ubi-minimal:latest` is the base image

nikhil-thomas replied:
done

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+            "name": "example-install",
+            "namespace": "tekton-pipelines"
+        },
+          "spec": {}
```
Should the `spec` field have some configuration for the resulting application?

nikhil-thomas replied:
In the current version, we do not support any configuration. The install process is kept simple. We are planning to add spec fields in the future versions

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+        }
+      ]
+  name: tektoncd-pipeline-operator.v0.0.1
+  namespace: tekton-pipelines
```
nit: Please use `placeholder` for the operator namespace

nikhil-thomas replied:
done

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+      - description: Version of Tektoncd Pipeline installed
+        displayName: Tektoncd Pipeline Version
+        path: version
+  description: Kubernetes operator to install TektonCD-Pipeline
```
FYI: This description can be longer and multiline is supported in this case.

nikhil-thomas replied:
we have updated it in the latest commit

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,250 @@
+                - name: WATCH_NAMESPACE
+                  valueFrom:
+                    fieldRef:
+                      fieldPath: metadata.namespace
```
Please set `fieldPath`  to `metadata.annotations['olm.targetNamespaces']` so that OLM can inform the operator which namespace(s) it should watch

nikhil-thomas replied:
done

nikhil-thomas replied:
done

SamiSousa commented:
@nikhil-thomas If this has been tested on vanilla Kube, please also submit a copy of these manifests in the `upstream-community-operators` directory as well.

nikhil-thomas commented:
> @nikhil-thomas Thank you for submitting this operator! Is this operator's image built by the Red Hat build process?

i used `operator-sdk build`.
`registry.access.redhat.com/ubi8-dev-preview/ubi-minimal:latest` is the base image.

siamaksade commented:
/cc @pmorie @bbrowning

nikhil-thomas commented:
> @nikhil-thomas If this has been tested on vanilla Kube, please also submit a copy of these manifests in the `upstream-community-operators` directory as well.

done

dmesser commented:
Excited to try this one out!
Meanwhile @nikhil-thomas you could use [https://operatorhub.io/preview](https://operatorhub.io/preview) to preview how your CSV would render on the community site. I think it would be really good to get a more detailed description about the the Operators capability, the nature of the managed application (some info on Tekton CD) and any requireiments the user has to fulfil before using this Operator. All this would go into `spec.description`.

Thanks!

SamiSousa commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+  displayName: TektonCD Pipeline Operator
+  install:
+    spec:
+      clusterPermissions:
```
Does this operator require clusterPermissions to function properly in a SingleNamespace installMode?
When using a `permissions` section in the CSV, if the CSV is installed with the AllNamespaces installMode, OLM will upgrade the permissions to clusterPermissions automatically.

nikhil-thomas replied:
I see. I shall test locally and update it

nikhil-thomas replied:
yes, it does. This operator creates cluster permissions for openshift-pipe lines. So taht openshift pipelines can watch `taskruns` `pipelineruns` etc created in any namespace.

At present `AllNamespaces` installMode doesn't fit our purpose. So we have set `AllNamespaces` installMode to false.

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+metadata:
+  annotations:
+    capabilities: Basic Install
+    description: An operator to install TektonCD Pipeline
```
The description should be `An operator to install OpenShift Pipelines`

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+  annotations:
+    capabilities: Basic Install
+    description: An operator to install TektonCD Pipeline
+    containerImage: quay.io/openshift-pipeline/nikthoma_tektoncd-pipeline-operator:v0.2.0-1
```
Shouldn't the image be `quay.io/openshift-pipeline/tektoncd-pipeline-operator`?

nikhil-thomas replied:
updated

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+          "apiVersion": "tekton.dev/v1alpha1",
+          "kind": "Install",
+          "metadata": {
+            "name": "example-install"
```
The `example-install` is what most users leave as the name. Should be rather something like `openshift-pipelines-install`.

nikhil-thomas replied:
@siamaksade I have merged your pr into my branch

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+  links:
+  - name: Tekton Pipeline GitHub Repository
+    url: https://github.com/tektoncd/pipeline
+  - name: Tekton Pipeline Operator GitHub Repository
```
`OpenShift Pipelines Operator GitHub Repository` since this is the downstream operator

nikhil-thomas replied:
done

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+    - name: installs.tekton.dev
+      version: v1alpha1
+      kind: Install
+      displayName: TektonCD-Pipeline Install
```
`displayName` should refer to the downstream name. `OpenShift Pipelines Install`

SamiSousa replied:
Should the operator packageName also match this "downstream name"?

nikhil-thomas replied:
@siamaksade @SamiSousa 
the displayname has been updated. I shall update the packagename as well

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+      version: v1alpha1
+      kind: Install
+      displayName: TektonCD-Pipeline Install
+      description: CRD to trigger install/delete of TektonCD-Pipeline
```
Should be updated to downstream name

siamaksade commented on upstream-community-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+                      fieldPath: metadata.name
+                - name: OPERATOR_NAME
+                  value: tektoncd-pipeline-operator
+                image: quay.io/openshift-pipeline/nikthoma_tektoncd-pipeline-operator:v0.2.0-1
```
Update image ref to `openshift-pipeline/tektoncd-pipeline-operator`

nikhil-thomas replied:
updated:
quay.io/openshift-pipeline/openshift-pipelines-operator

siamaksade commented on redhat-operators/tektoncd-pipeline-operator/tektoncd-pipeline-operator.v0.0.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,248 @@
+  name: tektoncd-pipeline-operator.v0.0.1
+  namespace: placeholder
+spec:
+  version: 0.0.1
```
How does this version correlate with the operator image version?

SamiSousa replied:
This CSV version need not match the image version, but may match if you so choose.
There are operators that choose to match image version with CSV version as a sort of convention.

nikhil-thomas replied:
i think it will be better if the csv version match the image version.
I have updated it.

nikhil-thomas commented:
> Excited to try this one out!
> Meanwhile @nikhil-thomas you could use https://operatorhub.io/preview to preview how your CSV would render on the community site. I think it would be really good to get a more detailed description about the the Operators capability, the nature of the managed application (some info on Tekton CD) and any requireiments the user has to fulfil before using this Operator. All this would go into `spec.description`.
> 
> Thanks!

Thank you for the comment. We have updated the description.
At present the user does not have to fulfill any prerequisites.

hrishin commented on redhat-operators/openshift-pipelines-operator/openshift-pipelines-operator-tekton_v1alpha1_install_crd.yaml:
```diff
@@ -0,0 +1,49 @@
+              type: array
+            version:
+              type: string
+          required:
```
Wondering, what could happen if it's not marked as required? I was under impression `required` apply constraint during resource creation/update, right?

nikhil-thomas replied:
Right now we don't have any spec defined.
That required is in status which is handled by the Operator so the user will not have to worry about this required statement

nikhil-thomas replied:
Right now we don't have any spec defined.
That required is in status which is handled by the Operator so the user will not have to worry about this required statement.

hrishin commented on upstream-community-operators/openshift-pipelines-operator/openshift-pipelines-operator.v0.3.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,285 @@
+    capabilities: Basic Install
+    categories: "Developer Tools"
+    description: OpenShift Pipelines is a cloud-native CI/CD solution for building pipelines using Tekton concepts which run natively on OpenShift and Kubernetes.
+    containerImage: quay.io/openshift-pipeline/openshift-pipelines-operator:v0.3.1-1
```
If it's upstream(assuming the tekton) then shouldn't it refer tekton community rather RedHat? This image is the same as 
[downstream](https://github.com/operator-framework/community-operators/pull/296/files#diff-4cb1ff32039ec21b3cea67beb7c4a989R8) one, so wondering :thinking: why we need to repeat the same yaml manifests twice?

nikhil-thomas replied:
Both of these CSV s represents the downstream operator

we added it twice (in two different directories) for two different platforms
refer: https://github.com/operator-framework/community-operators/blob/master/README.md#where-to-contribute

SamiSousa commented:
/lgtm
Putting this PR on hold until after the freeze for OpenShift operators ends
/hold

nikhil-thomas commented:
> /lgtm
> Putting this PR on hold until after the freeze for OpenShift operators ends
> /hold

Thanks :+1:

cvp-ops commented:
PR includes changes to only the [upstream-community-operators] directory. No pipeline will be launched.

nikhil-thomas commented:
@SamiSousa
i added and icon and updated the description today
there is a check failing: https://github.com/nmars/cvp-community-operators/blob/master/community-operators/COMMUNITY-OPERATORS-PULL-REQUESTS.md

nikhil-thomas commented:
I have split this PR into 2 one for redhat-operators and one for upstream-community-operators
https://github.com/operator-framework/community-operators/pull/354
https://github.com/operator-framework/community-operators/pull/355

if we merge those, then we can close this pr.

nikhil-thomas commented:
openshift-pipelines-operator got merged into `community-operators` directory in #362
Hence, closing this pr.
